### ISSUE TITLE
recent update causes gcc to be unusable

### ISSUE THREAD

**Issue by trws** [LLNL/spack#1244]
After a recent pull, spack fails to use gcc.  It either claims I'm on an operating system I'm not on (which is interesting in and of itself) or attempts to load a module that doesn't exist.  The latter problem appears to be caused by this change: https://github.com/LLNL/spack/blame/3f90fee614cf44f31344de3443f2a2e63c6d9fb4/lib/spack/spack/compilers/gcc.py#L52

**Comment by trws** [comment]
It seems to be a conflict between different operating system detection methods, where a cray system is not running cray's full userland.

**Comment by tgamblin** [comment]
@becker33: can you help Tom out?

**Comment by tgamblin** [comment]
@trws: which cray machine is this?

**Comment by trws** [comment]
rzhasgpu, cray machine, cray compiler modules available for cray compilers, TOSS2 userland

**Comment by becker33** [comment]
@twrs: What is the architecture spack claims you are running on? Is it cray_xc-SuSE11-x86_64?

-Greg

**Comment by trws** [comment]
cray_xc-CNL10-ivybridge

**Comment by trws** [comment]
module error:

```
scogland at rzhasgpu18 in ~
$ spack install the_platinum_searcher
==> Installing the_platinum_searcher
==> Installing go
==> Installing go-bootstrap
==> Installing git
==> Installing autoconf
==> Installing m4
==> Installing libsigsegv
Traceback (most recent call last):
  File "/usr/workspace/wsrzc/scogland/spack/bin/spack", line 179, in <module>
    main()
  File "/usr/workspace/wsrzc/scogland/spack/bin/spack", line 157, in main
    return_val = command(parser, args)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/cmd/install.py", line 91, in install
    explicit=True)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 940, in do_install
    run_tests=run_tests)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 1093, in do_install_dependencies
    dep.package.do_install(**kwargs)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 940, in do_install
    run_tests=run_tests)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 1093, in do_install_dependencies
    dep.package.do_install(**kwargs)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 940, in do_install
    run_tests=run_tests)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 1093, in do_install_dependencies
    dep.package.do_install(**kwargs)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 940, in do_install
    run_tests=run_tests)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 1093, in do_install_dependencies
    dep.package.do_install(**kwargs)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 940, in do_install
    run_tests=run_tests)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 1093, in do_install_dependencies
    dep.package.do_install(**kwargs)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 940, in do_install
    run_tests=run_tests)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 1093, in do_install_dependencies
    dep.package.do_install(**kwargs)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/package.py", line 1050, in do_install
    spack.build_environment.fork(self, build_process, dirty=dirty)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/build_environment.py", line 509, in fork
    setup_package(pkg, dirty=dirty)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/build_environment.py", line 447, in setup_package
    set_compiler_environment_variables(pkg, spack_env)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/build_environment.py", line 222, in set_compiler_environment_variables
    load_module(mod)
  File "/usr/workspace/wsrzc/scogland/spack/lib/spack/spack/build_environment.py", line 137, in load_module
    exec(compile(load, '<string>', 'exec'))
  File "<string>", line 1
    ModuleCmd_Load.c(208):ERROR:105: Unable to locate a modulefile for 'PrgEnv-gnu'
                         ^
SyntaxError: invalid syntax
==> Error: Installation process had nonzero exit code : 256
```

**Comment by trws** [comment]
Having worked around that, by rewriting the compilers.yaml to actually use the compilers correctly, I get errors telling me that no version of a package matches ":".  Something is very thoroughly broken here...

**Comment by becker33** [comment]
@twrs: The issue is with how spack detects cray machines and chooses defaults. It’s trying to build for the backend cray environment.

If you make the following changes in spack/lib/spack/spack/platforms/cray_xc.py it should work. Replace “SuSE11” with “redhat6” everywhere it occurs. This replaces the cray front-end operating system. Then change `default_os = “CNL10”` to `default_os = “redhat6”`. This will make spack choose the front end operating system by default. You will then want to specify “os=be” or “os=backend” if you want to use the cray compilers.

I don’t think there are many use cases for these sorts of unusual cray machines outside of rzhasgpu, so I’m not sure it’s worth pushing any auto detection for this back into spack. If it comes up at other sites we should look into it.

-Greg

**Comment by trws** [comment]
@becker33, how is it detecting that it's a cray environment, and why would that cause all of these issues?  The system is binary compatible with all of the other TOSS2 systems, it's linux, on intel, and the entire package manager is broken because it happens to have some cray modules?  This seems... unreasonable.  I agree that there aren't likely to be a lot of cray systems around like this, but are there systems that could be mis-identified as using a cray userland like this?

**Comment by becker33** [comment]
@trws: Spack currently detects the cray_xc platform by looking for the /opt/cray/craype directory. We have thought about changing this, but there are not very good ways to identify cray machines.

If anyone has a suggestion for how better to identify cray systems, we would certainly welcome it.

-Greg

**Comment by trws** [comment]
Perhaps at least see if the standard gnu environment module exists 
before determining that it is a cray system? That's one of the big 
issues for me right now, in that it emits requests for all of these 
modules that do not exist.

**Comment by becker33** [comment]
@trws: @tgamblin: The directory /cray_home appears to be used only in the implementation of the cray environment. I’ve submitted PR #1245 to change our detection method to key on the /cray_home directory.

-Greg

**Comment by trws** [comment]
Seems plausible.  @tgamblin?

**Comment by becker33** [comment]
Closing this as resolved via #1245
